subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileToolchainIntegrationTest.groovy

```diff
@@ -469,6 +469,55 @@ class JavaCompileToolchainIntegrationTest extends AbstractIntegrationSpec implem
         JavaVersion.current()   | "[deprecation] foo() in Foo has been deprecated"
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/23990")
+    @Requires(TestPrecondition.JDK11_OR_LATER)
+    def "can compile with a custom compiler executable"() {
+        def jdk = AvailableJavaHomes.getJdk(JavaVersion.current())
+
+        buildFile << """
+            plugins {
+                id("java")
+            }
+
+            configurations {
+                ecj {
+                    canBeConsumed = false
+                    canBeResolved = true
+                }
+            }
+
+            ${mavenCentralRepository()}
+
+            dependencies {
+                ecj("org.eclipse.jdt:ecj:3.32.0")
+            }
+
+            compileJava {
+                def customJavaLauncher = javaToolchains.launcherFor {
+                    languageVersion.set(JavaLanguageVersion.of(${jdk.javaVersion.majorVersion}))
+                }.get()
```

Comment:
I assume that this would still fail if a different default toolchain is configured at project level. `javaCompiler` would use `javac` from the default toolchain and the selected toolchain here might not necessarily match. It would probably be a good idea to also test that situation (which I think should be supported too).

Reply:
I guess this could be worked around by retrieving the launcher for the toolchain of the configured `JavaCompiler` somehow but I believe that getting the toolchain for a tool is not currently possible.

Reply:
> I assume that this would still fail if a different default toolchain is configured at project level.

It should not fail. Since we are configuring `forkOptions.executable` for this `JavaCompile` task, it will take priority over any toolchain configured on the project level. Here is how the override is selected:
https://github.com/gradle/gradle/blob/c11d14a66cf819d659e242760f430288f47ff028/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java#L105-L108

I'll update the test to show this